### PR TITLE
Show transformations in the ellipsis menu

### LINKED ISSUE #3434: Show transformations in the ellipsis menu
## Issue Overview

We've recently added "convert to blocks" as an option to the block level ellipsis menu on the Classic Text block. This feels like a natural place for this, since it's an advanced transformation at the block level.

Should we show other transforms in this menu? This would be a pattern that we could reuse from the current "Switcher" interface. 

## Mockups

When selecting on a single block:

single select transform

When multiple blocks are selected:

multi select transform

In these mockups we show _the most obvious transformation directly_ (that would be gallery when selecting multiple images, or perhaps quote when selecting multiple text blocks), as well as a "Transform into" popout menu. 

Is this a good design? Is the popout menu too complex? Do we think it might be a scalable pattern to show all transformations directly in the menu?

Perhaps this issue could be tackled in phases, to test what works.

### PR DESCRIPTION
**Pull request by jorgefilipecosta**
## Description
This PR aims to close https://github.com/WordPress/gutenberg/issues/3434 by adding the transformations to the block side menu.

The getPossibleBlockTransformations function was extracted from BlockSwitcher component as it was in order for us to be able to use this logic outside of BlockSwitcher component.

## How Has This Been Tested?
Add some blocks e.g: paragraph, list, heading press the side menu, verify the transformations appear. Apply the transformations and see the transformations happened as expected.
Select a sequence of paragraphs, transform them into a list. Transform the list back into paragraphs.
Add a gallery, transform into images, select all the images and transform into a gallery.

## Screenshots (jpeg or gifs if applicable):
screen shot 2017-11-14 at 18 10 39
screen shot 2017-11-14 at 18 11 30

### CODE REVIEW

**Review by karmatosed — CHANGES_REQUESTED** [review]
In the mockups there is a separator splitting out the menu items. This visual indicator I think was really useful. Could we add it back in?

**Comment by jasmussen** [comment]
Nice work, thanks for working on this. I think this'll be really nice to have — it feels natural for these to be "properties of a multi selection".

In addition to Tammies suggestion, can you move the transformations below the delete button? That way we group transformations all at the bottom.

File: blocks/api/factory.js [hunk 1]
```diff
@@ -67,6 +68,48 @@ export function createBlock( name, blockAttributes = {} ) {
 }
 
 /**
+ * Returns an array of possible block transformations that could happen on the set of blocks received as argument.
+ *
+ * @param  {Array}  blocks Blocks array
+ * @return {Array}         Array of possible block transformations
+ */
+export function getPossibleBlockTransformations( blocks ) {
```
Comment (gziolo) [inline blocks/api/factory.js]: Since it is nicely extracted into its own utility method, it would be even better to cover this functionality with unit tests :)

**Comment by gziolo** [comment]
I think the idea here is to remove transformed from the top toolbar:
screen shot 2017-11-14 at 09 37 28

@jasmussen, can you confirm?

**Comment by gziolo** [comment]
I like this change a lot, nice work so far 💯

**Comment by youknowriad** [comment]
Should we remove the existing block switchers from the Block's toolbar? Same for multi-selection.

**Comment by jasmussen** [comment]
> I think the idea here is to remove transformed from the top toolbar:

_Maybe_. It depends a little bit.

Right now we default to a top-docked toolbar. The benefit of this is that it's always there in the same spot, never covering the content. The downside is that in some cases, notably around non-text blocks, it can feel detached from the content, compared to the block-docked toolbar, which you can switch to in the ellipsis menu.

We mean to test these two on a bunch of people, see which works the best for the most people. This might define which version we default to, though which ever side we land on it'd be nice to keep the switch around.

The thing is, in the case of the top-docked toolbar, we can do something like this:

direct transformations

That is, have direct transformations linked right in the top toolbar, making it not only contextual to the block selected but indeed the _blocks_.

We can't really do that with block-docked toolbars, we've explored that in the past and due to the nature of how `position: sticky` works, we'd have to wrap the selected blocks in a new div on the fly, in order for the toolbar to be able to scroll with you down the page. Complex, slightly hacky, and has its own sets of problems.

So, long story short — there are _opportunities_ in that top area for having direct transformations there. But it depends a little bit on how things play out in tests.

Did that make sense?

**Comment by gziolo** [comment]
> That is, have direct transformations linked right in the top toolbar, making it not only contextual to the block selected but indeed the blocks.

This might be a very interesting pattern when multiple blocks are selected. It's also an improved version of what we have at the moment:

screen shot 2017-11-14 at 10 25 21

> So, long story short — there are opportunities in that top area for having direct transformations there. But it depends a little bit on how things play out in tests.

I was a little bit vague. My point was, that there is some duplication with the current implementation which this PR proposes. I think we can keep both versions for the time being and decide later if that is helpful. Code-wise it will be a very straightforward task to adjust it later 👍

**Comment by jasmussen** [comment]
> My point was, that there is some duplication with the current implementation which this PR proposes. I think we can keep both versions for the time being and decide later if that is helpful. Code-wise it will be a very straightforward task to adjust it later 👍

Right, I have no strong opinion on having both for the time being, or removing the top one. Just that depending on how the top toolbar performs, we might want to revisit that.

**Comment by jorgefilipecosta** [comment]
Hi @karmatosed, thank you for your review, some changes were applied to get closer to the matchups (border added, transformations moved to be after delete), feel free to have a new look.

Hi @gziolo, @jasmussen, regarding the duplication of the transforms. Removing them from top bar is very simple and can be easily done. If you prefer I can also create a parallel pull request that shows the direct transformations.

**Comment by gziolo** [comment]
> Hi @gziolo, @jasmussen, regarding the duplication of the transforms. Removing them from top bar is very simple and can be easily done. If you prefer I can also create a parallel pull request that shows the direct transformations.

I would trust Joen and Tammie in this regard :) I'm fine with all options, just wanted to make it clear that we have the same options available in two places.

File: blocks/api/factory.js [hunk 2]
```diff
@@ -67,6 +68,48 @@ export function createBlock( name, blockAttributes = {} ) {
...
+	if ( ! blocks || ! blocks[ 0 ] ) {
+		return [];
+	}
+	const isMultiBlock = blocks.length > 1;
+	const sourceBlockName = blocks[ 0 ].name;
+
+	if ( isMultiBlock && ! every( blocks, ( block ) => ( block.name === sourceBlockName ) ) ) {
```
Comment (aduth) [inline blocks/api/factory.js]: Minor: Most Lodash methods allow predicates to be defined using an object shorthand. In this case:

```js
// Before
! every( blocks, ( block ) => ( block.name === sourceBlockName ) )

// After
! every( blocks, { name: sourceBlockName } )
```

The only potential worry is that it is listed as being removed by default in the 5.0.0 roadmap:

https://github.com/lodash/lodash/wiki/Roadmap

File: blocks/api/factory.js [hunk 3]
```diff
@@ -67,6 +68,48 @@ export function createBlock( name, blockAttributes = {} ) {
 }
 
 /**
+ * Returns an array of possible block transformations that could happen on the set of blocks received as argument.
+ *
+ * @param  {Array}  blocks Blocks array
+ * @return {Array}         Array of possible block transformations
+ */
+export function getPossibleBlockTransformations( blocks ) {
+	if ( ! blocks || ! blocks[ 0 ] ) {
```
Comment (aduth) [inline blocks/api/factory.js]: Is this meant to be testing for a falsey or empty array? If so, I would suggest at least testing `Array#length` instead of accessing the first member directly. Alternatively consider [`_.isEmpty`](https://lodash.com/docs/4.17.4#isEmpty).

_Edit:_ Okay, I see we're working on the first entry in the set. Maybe we could assign `sourceBlock` as `first( blocks )` and test its truthiness here?

https://lodash.com/docs/4.17.4#first

File: blocks/api/factory.js [hunk 4]
```diff
@@ -67,6 +68,48 @@ export function createBlock( name, blockAttributes = {} ) {
...
+			transformFrom,
+			t => t.type === 'block' && t.blocks.indexOf( sourceBlockName ) !== -1 &&
+				( ! isMultiBlock || t.isMultiBlock )
+		);
+		return transformation ? memo.concat( [ type.name ] ) : memo;
+	}, [] );
+	const blocksToBeTransformedTo = get( blockType, 'transforms.to', [] )
+		.reduce(
+			( memo, transformation ) =>
+				memo.concat( ! isMultiBlock || transformation.isMultiBlock ? transformation.blocks : [] ),
```
Comment (aduth) [inline blocks/api/factory.js]: Can't we just assign `blocksToBeTransformedTo` to `blockType.transforms.to` if `isMultiBlock` and skip the `Array#reduce` for those cases where it's constant?

**Comment by jorgefilipecosta** [comment]
Hi @aduth, thank you for your review. In fact, I end up applying a huge refactor to getPossibleBlockTransformations (function extracted from block-switcher). I replace all reduce+concat occurrences with flatMap, compact+map, or filter+map. I think this makes the code easier to understand and more performant. Feel free to have a new look.

File: blocks/api/factory.js [hunk 5]
```diff
@@ -62,6 +67,44 @@ export function createBlock( name, blockAttributes = {} ) {
...
+
+	if ( isMultiBlock && ! every( blocks, { name: sourceBlockName } ) ) {
+		return [];
+	}
+
+	const blocksToBeTransformedFrom = filter( getBlockTypes(), type =>
+		find(
+			get( type, 'transforms.from', [] ),
+			t => t.type === 'block' && t.blocks.indexOf( sourceBlockName ) !== -1 &&
+			( ! isMultiBlock || t.isMultiBlock )
```
Comment (aduth) [inline blocks/api/factory.js]: It's hard to tell that this is part of the predicate for `find`, especially since the two lines above are separate arguments. I would suggest one of:

- Indent one more to indicate continuation of the condition
- Be more explicit with function arguments and avoid the arrow implicit return syntax

```js
const blocksToBeTransformedFrom = filter( getBlockTypes(), ( type ) => {
	return find( get( type, 'transforms.from', [] ), ( transform ) => {
		return (
			transform.type === 'block' &&
			transform.blocks.indexOf( sourceBlockName ) !== -1 &&
			( ! isMultiBlock || transform.isMultiBlock )
		);
	} );
} ).map( type => type.name );
```

(Note I also tend to discourage abbreviations. In this case it is very difficult to determine what `t` is - type? transform?)

- Split out separate named functions which better explain each step along the way

```js
const isTransformForBlockSource = ( sourceName ) => ( transform ) => (
	transform.type === 'block' &&
	transform.blocks.indexOf( sourceName ) !== -1 &&
	( ! isMultiBlock || transform.isMultiBlock )
);

const createIsTypeTransformableFrom = ( sourceName ) => ( type ) => (
	!! find(
		get( type, 'transforms.from', [] ),
		isTransformForBlockSource( sourceName ),
	)
);

const blocksToBeTransformedFrom = filter(
	getBlockTypes(),
	createIsTypeTransformableFrom( sourceBlockName ),
).map( type => type.name );
```

I think some inline comments (and/or named functions) would really go a long way to improve readability here. It's quite difficult to determine what's going on.

File: blocks/api/factory.js (hunk 1 above)
Comment (aduth) [reply blocks/api/factory.js]: >Since it is nicely extracted into its own utility method, it would be even better to cover this functionality with unit tests :)

Agree with this. Can we get some tests for this?

File: blocks/api/factory.js [hunk 6]
```diff
@@ -62,6 +67,44 @@ export function createBlock( name, blockAttributes = {} ) {
...
+	).map( type => type.name );
+
+	const blockType = getBlockType( sourceBlockName );
+	const transformsTo = get( blockType, 'transforms.to', [] );
+	const blocksToBeTransformedTo = flatMap(
+		isMultiBlock ? filter( transformsTo, transformation => transformation.isMultiBlock ) : transformsTo,
+		transformation => transformation.blocks
+	);
+
+	return uniq( compact(
```
Comment (aduth) [inline blocks/api/factory.js]: There's some benefit here to clarify intent, at least with the unique (we expected duplicates, we expect some falsey types), but it seems this could be simplified to a reduce where we check for those within the iterator of the reduce:

```js
return reduce( [
	...blocksToBeTransformedFrom,
	...blocksToBeTransformedTo,
], ( result, name ) => {
	const transformBlockType = getBlockType( name );
	if ( transformBlockType && ! includes( result, transformBlockType ) ) {
		result.push( transformBlockType );
	}

	return result;
}, [] );
```

Comment (aduth) [reply blocks/api/factory.js]: >There's some benefit here to clarify intent, at least with the unique (we expected duplicates, we expect some falsey types), but it seems this could be simplified to a reduce where we check for those within the iterator of the reduce:

Maybe worth pointing out that this might be more of a personal preference thing, and a desire to reduce the number of iterating operations, and that others might find your original approach to be clearer. I think the spread syntax could be a nice touch at least to clarify the joined set.

**Comment by jorgefilipecosta** [comment]
Hi @aduth, @gziolo, this PR was updated with test cases added to the extracted function. @aduth, thank you for providing deep inputs about how to make the function easier to understand I used most of them.

File: editor/components/block-switcher/index.js [hunk 7]
```diff
@@ -25,41 +24,14 @@ import { getBlock } from '../../selectors';
...
-	const sourceBlockName = blocks[ 0 ].name;
+	const allowedBlocks = getPossibleBlockTransformations( blocks );
 
-	if ( isMultiBlock && ! every( blocks, ( block ) => ( block.name === sourceBlockName ) ) ) {
+	if ( ! allowedBlocks.length ) {
 		return null;
 	}
 
+	const sourceBlockName = blocks[ 0 ].name;
 	const blockType = getBlockType( sourceBlockName );
```
Comment (aduth) [inline editor/components/block-switcher/index.js]: Kinda makes me wish that `<BlockIcon />` could render via the name of the block, so we wouldn't have to get the full type ahead of type in usage.

File: blocks/api/factory.js [hunk 8]
```diff
@@ -56,6 +60,77 @@ export function createBlock( name, blockAttributes = {} ) {
...
+
+	const blocksToBeTransformedFrom = filter(
+		getBlockTypes(),
+		createIsTypeTransformableFrom( sourceBlockName, isMultiBlock ),
+	).map( type => type.name );
+
+	const blockType = getBlockType( sourceBlockName );
+	const transformsTo = get( blockType, 'transforms.to', [] );
+	const blocksToBeTransformedTo = flatMap(
+		isMultiBlock ? filter( transformsTo, transformation => transformation.isMultiBlock ) : transformsTo,
```
Comment (aduth) [inline blocks/api/factory.js]: Minor: There's a shorthand for this `filter` usage:

```js
filter( transformsTo, 'isMultiBlock' )
```

**Review by aduth — APPROVED** [review]
Seems to work nicely, and splitting out the transformation logic seems to lend some more clarity. Still might do for a few inline comments on `getPossibleBlockTransformations`, but generally 👍

File: editor/components/block-switcher/index.js (hunk 7 above)
Comment (jorgefilipecosta) [reply editor/components/block-switcher/index.js]: Yes, in most cases we have the block, so it makes sense to receive icon but it would make sense to also be possible to pass the name of the block. I will propose this change in a separate PR.

File: editor/components/block-settings-menu/block-transformations.js [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+/**
+ * Internal dependencies
+ */
+import './style.scss';
+import { getBlock } from '../../selectors';
+import { replaceBlocks } from '../../actions';
+
+function BlockTransformations( { blocks, small = false, onTransform, onClick = noop } ) {
+	return <div className="editor-block-settings-menu__block-transformations">
+		{ getPossibleBlockTransformations( blocks ).map( ( { name, title, icon } ) => {
```
[the remark this answers is not in the document]
Comment (jorgefilipecosta) [reply editor/components/block-settings-menu/block-transformations.js]: Given that switchToBlockType receives an array of blocks (and needs to) it may make sense that getPossibleBlockTransformations receive the same array. The places that use one function also use the other so if we use an array of slugs we will probably just add a map to create an array of slugs from the array of blocks and then inside the function, a map in the opposite direction. Also receiving an array blocks leaves the possibility in the future for transformations to depend on the presence of some attribute or another more advanced use case.

**Comment by jorgefilipecosta** [comment]
Hi @karmatosed, could you check if the design changes here are ready to ship :)?

File: blocks/api/factory.js (hunk 1 above)
Comment (jorgefilipecosta) [reply blocks/api/factory.js]: tests were added :)

**Comment by karmatosed** [comment]
It seems we lost the version beside the actual block in the latest version?

2017-12-14 at 12 47

Was that intended @jasmussen? Just reading through this I don't think it was.

**Comment by jasmussen** [comment]
This is looking good to me! 

One thing, though — it looks like the separator that separates Settings/Delete from transformations is black. Can you make that one `$light-gray-500`?

screen shot 2017-12-14 at 14 31 08

**Comment by jorgefilipecosta** [comment]
Hi @jasmussen, nice catch, the change was applied.

**Comment by karmatosed** [comment]
All looks good to me now.. lets ship this! Thanks for working through.

**Review by karmatosed — APPROVED** [review]
